### PR TITLE
GUAC-1389: Refactor libguac to allow screen sharing.

### PR DESCRIPTION
**Pull request by mike-jumper**
Parts of the build that will break have been disabled temporarily.

### CODE REVIEW

File: src/libguac/client.c [hunk 1]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
 }
 
-guac_object* guac_client_alloc_object(guac_client* client) {
 
-    guac_object* allocd_object;
-    int object_index;
+/**
+ * The broadcast socket cannot be read from.
+ */
+static ssize_t __guac_socket_broadcast_read_handler(guac_socket* socket,
```
Comment (jmuehlner) [inline src/libguac/client.c]: Missing some parameter comments here.

File: src/libguac/client.c [hunk 2]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
-        return NULL;
+    /* Broadcast socket reads are not allowed */
+    return -1;
+
+}
+
+/**
+ * Writes a chunk of data to a given user.
+ */
+static void* __write_chunk_callback(guac_user* user, void* data) {
```
Comment (jmuehlner) [inline src/libguac/client.c]: Missing some more  parameter comments here.

File: src/libguac/client.c [hunk 3]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
+    return NULL;
 
 }
 
-void guac_client_free_object(guac_client* client, guac_object* object) {
+/**
+ * Socket write handler which operates on each of the sockets of all connected
+ * users, unifying the results.
+ */
+static ssize_t __guac_socket_broadcast_write_handler(guac_socket* socket,
```
Comment (jmuehlner) [inline src/libguac/client.c]: Missing comments

File: src/libguac/client.c [hunk 4]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
 }
 
-/**
- * Returns a newly allocated string containing a guaranteed-unique connection
- * identifier string which is 37 characters long and begins with a '$'
- * character.  If an error occurs, NULL is returned, and no memory is
- * allocated.
- */
-static char* __guac_generate_connection_id() {
+static void* __flush_callback(guac_user* user, void* data) {
```
Comment (jmuehlner) [inline src/libguac/client.c]: Comment?

File: src/libguac/client.c [hunk 5]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
+}
 
-    /* Generate random UUID */
-    if (uuid_make(uuid, UUID_MAKE_V4) != UUID_RC_OK) {
-        uuid_destroy(uuid);
-        guac_error = GUAC_STATUS_NO_MEMORY;
-        guac_error_message = "UUID generation failed";
-        return NULL;
-    }
+static ssize_t __guac_socket_broadcast_flush_handler(guac_socket* socket) {
```
Comment (jmuehlner) [inline src/libguac/client.c]: Missing comment

File: src/libguac/client.c [hunk 6]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
-        guac_error = GUAC_STATUS_INTERNAL_ERROR;
-        guac_error_message = "Conversion of UUID to connection ID failed";
-        return NULL;
-    }
+    return 0;
 
-    uuid_destroy(uuid);
+}
+
+static void* __lock_callback(guac_user* user, void* data) {
```
Comment (jmuehlner) [inline src/libguac/client.c]: Missing comment

File: src/libguac/client.c [hunk 7]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
+
+    /* Lock socket */
+    guac_socket_instruction_begin(user->socket);
+
+    return NULL;
+
+}
+
+static void __guac_socket_broadcast_lock_handler(guac_socket* socket) {
```
Comment (jmuehlner) [inline src/libguac/client.c]: Missing comment

File: src/libguac/client.c [hunk 8]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
+
+    /* Flush all users */
+    guac_client_foreach_user(client, __lock_callback, NULL);
+
+}
 
-    buffer[0] = '$';
-    buffer[UUID_LEN_STR + 1] = '\0';
-    return buffer;
+static void* __unlock_callback(guac_user* user, void* data) {
```
Comment (jmuehlner) [inline src/libguac/client.c]: Missing comment

File: src/libguac/client.c [hunk 9]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
+
+    /* Lock socket */
+    guac_socket_instruction_end(user->socket);
+
+    return NULL;
+
+}
+
+static void __guac_socket_broadcast_unlock_handler(guac_socket* socket) {
```
Comment (jmuehlner) [inline src/libguac/client.c]: Missing comment

File: src/libguac/client.c [hunk 10]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
+    guac_client_foreach_user(client, __lock_callback, NULL);
+
+}
 
-    buffer[0] = '$';
-    buffer[UUID_LEN_STR + 1] = '\0';
-    return buffer;
+static void* __unlock_callback(guac_user* user, void* data) {
+
+    /* Lock socket */
```
Comment (jmuehlner) [inline src/libguac/client.c]: Unlock?

File: src/libguac/client.c [hunk 11]
```diff
@@ -124,107 +134,132 @@ guac_stream* guac_client_alloc_stream(guac_client* client) {
...
+
+    /* Flush all users */
+    guac_client_foreach_user(client, __unlock_callback, NULL);
+
+}
+
+/**
+ * The broadcast socket cannot be read from (nor selected).
+ */
+static int __guac_socket_broadcast_select_handler(guac_socket* socket, int usec_timeout) {
```
Comment (jmuehlner) [inline src/libguac/client.c]: Missing parameter comments

File: src/libguac/client.c [hunk 12]
```diff
@@ -381,6 +409,198 @@ void guac_client_abort(guac_client* client, guac_protocol_status status,
...
+int guac_client_load_plugin(guac_client* client, const char* protocol) {
+
+    /* Reference to dlopen()'d plugin */
+    void* client_plugin_handle;
+
+    /* Pluggable client */
+    char protocol_lib[GUAC_PROTOCOL_LIBRARY_LIMIT] =
+        GUAC_PROTOCOL_LIBRARY_PREFIX;
+ 
+    union {
```
Comment (jmuehlner) [inline src/libguac/client.c]: Maybe explain this a bit?

File: src/libguac/guacamole/client-fntypes.h [hunk 13]
```diff
@@ -34,102 +34,25 @@
...
  * Handler for freeing up any extra data allocated by the client
  * implementation.
  */
 typedef int guac_client_free_handler(guac_client* client);
 
 /**
- * Handler for logging messages
+ * Handler for logging messages.
  */
 typedef void guac_client_log_handler(guac_client* client, guac_client_log_level level, const char* format, va_list args);
```
Comment (jmuehlner) [inline src/libguac/guacamole/client-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/client-fntypes.h [hunk 14]
```diff
@@ -34,102 +34,25 @@
...
- * Handler for logging messages
+ * Handler for logging messages.
  */
 typedef void guac_client_log_handler(guac_client* client, guac_client_log_level level, const char* format, va_list args); 
 
 /**
  * Handler which should initialize the given guac_client.
  */
-typedef int guac_client_init_handler(guac_client* client, int argc, char** argv);
+typedef int guac_client_init_handler(guac_client* client);
```
Comment (jmuehlner) [inline src/libguac/guacamole/client-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/client-fntypes.h [hunk 15]
```diff
@@ -34,102 +34,25 @@
...
-
-/**
  * Handler for freeing up any extra data allocated by the client
  * implementation.
  */
 typedef int guac_client_free_handler(guac_client* client);
 
 /**
- * Handler for logging messages
+ * Handler for logging messages.
```
Comment (jmuehlner) [inline src/libguac/guacamole/client-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/client.h [hunk 16]
```diff
@@ -435,33 +151,119 @@ struct guac_client {
...
+     *         "username",
+     *         "password",
+     *         NULL
+     *     };
+     *
+     *     int guac_client_init(guac_client* client) {
+     *         client->args = __my_args;
+     *     }
+     * @endcode
+
```
Comment (jmuehlner) [inline src/libguac/guacamole/client.h]: ?

File: src/libguac/guacamole/client.h [hunk 17]
```diff
@@ -717,10 +604,11 @@ void guac_client_stream_jpeg(guac_client* client, guac_socket* socket,
  * instruction) as WebP-encoded data at the given quality. The image stream
  * will be automatically allocated and freed. If the server does not support
  * WebP, this function has no effect, so be sure to check the result of
- * guac_client_supports_webp() prior to calling this function.
+ * guac_client_supports_webp() prior to calling
```
Comment (jmuehlner) [inline src/libguac/guacamole/client.h]: Why did you split this line?

File: src/libguac/guacamole/pool.h [hunk 18]
```diff
@@ -99,21 +106,29 @@ void guac_pool_free(guac_pool* pool);
...
+ *     The guac_pool to retrieve an integer from.
  *
- * @param pool The guac_pool to retrieve an integer from.
- * @return The next available integer, which may be either an integer not yet
- *         returned by a call to guac_pool_next_int, or an integer which was
- *         previosly returned, but has since been freed.
+ * @return
+ *     The next available integer, which may be either an integer not yet
+ *     returned by a call to guac_pool_next_int, or an integer which was
+ *     previosly returned, but has since been freed.
```
Comment (jmuehlner) [inline src/libguac/guacamole/pool.h]: Misspelling of "previously" here

File: src/libguac/guacamole/user-fntypes.h [hunk 19]
```diff
@@ -0,0 +1,140 @@
...
+#include "user-types.h"
+
+/**
+ * Callback which relates to a single guac_user at a time, along with arbitrary
+ * data.
+ *
+ * @see guac_client_foreach_user()
+ * @see guac_client_for_owner()
+ */
+typedef void* guac_user_callback(guac_user* user, void* data);
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 20]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Handler for Guacamole mouse events.
+ */
+typedef int guac_user_mouse_handler(guac_user* user, int x, int y,
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 21]
```diff
@@ -0,0 +1,140 @@
...
+        int button_mask);
+
+/**
+ * Handler for Guacamole key events.
+ */
+typedef int guac_user_key_handler(guac_user* user, int keysym, int pressed);
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 22]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Handler for Guacamole clipboard events.
+ */
+typedef int guac_user_clipboard_handler(guac_user* user, guac_stream* stream,
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 23]
```diff
@@ -0,0 +1,140 @@
...
+        char* mimetype);
+
+/**
+ * Handler for Guacamole screen size events.
+ */
+typedef int guac_user_size_handler(guac_user* user,
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 24]
```diff
@@ -0,0 +1,140 @@
...
+        int width, int height);
+
+/**
+ * Handler for Guacamole file transfer events.
+ */
+typedef int guac_user_file_handler(guac_user* user, guac_stream* stream,
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 25]
```diff
@@ -0,0 +1,140 @@
...
+        char* mimetype, char* filename);
+
+/**
+ * Handler for Guacamole pipe events.
+ */
+typedef int guac_user_pipe_handler(guac_user* user, guac_stream* stream,
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 26]
```diff
@@ -0,0 +1,140 @@
...
+        char* mimetype, char* name);
+
+/**
+ * Handler for Guacamole stream blob events.
+ */
+typedef int guac_user_blob_handler(guac_user* user, guac_stream* stream,
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 27]
```diff
@@ -0,0 +1,140 @@
...
+/**
+ * Handler for Guacamole stream ack events.
+ */
+typedef int guac_user_ack_handler(guac_user* user, guac_stream* stream,
+        char* error, guac_protocol_status status);
+
+/**
+ * Handler for Guacamole stream end events.
+ */
+typedef int guac_user_end_handler(guac_user* user, guac_stream* stream);
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 28]
```diff
@@ -0,0 +1,140 @@
...
+/**
+ * Handler for Guacamole stream blob events.
+ */
+typedef int guac_user_blob_handler(guac_user* user, guac_stream* stream,
+        void* data, int length);
+
+/**
+ * Handler for Guacamole stream ack events.
+ */
+typedef int guac_user_ack_handler(guac_user* user, guac_stream* stream,
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 29]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Handler for Guacamole stream end events.
+ */
+typedef int guac_user_end_handler(guac_user* user, guac_stream* stream);
+
+/**
+ * Handler for Guacamole audio format events.
+ */
+typedef int guac_user_audio_handler(guac_user* user, char* mimetype);
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 30]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Handler for Guacamole video format events.
+ */
+typedef int guac_user_video_handler(guac_user* user, char* mimetype);
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 31]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Handler for Guacamole join events. A join event is fired by the
+ * guac_client whenever a guac_user joins the connection.
+ */
+typedef int guac_user_join_handler(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 32]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Handler for Guacamole leave events. A leave event is fired by the
+ * guac_client whenever a guac_user leaves the connection.
+ */
+typedef int guac_user_leave_handler(guac_user* user);
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 33]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Handler for Guacamole sync events. A sync event is fired by the
+ * guac_client whenever a guac_user responds to a sync instruction.
+ */
+typedef int guac_user_sync_handler(guac_user* user, guac_timestamp timestamp);
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 34]
```diff
@@ -0,0 +1,140 @@
...
+
+/**
+ * Handler for Guacamole object get events.
+ */
+typedef int guac_user_get_handler(guac_user* user, guac_object* object,
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/guacamole/user-fntypes.h [hunk 35]
```diff
@@ -0,0 +1,140 @@
...
+        char* name);
+
+/**
+ * Handler for Guacamole object put events.
+ */
+typedef int guac_user_put_handler(guac_user* user, guac_object* object,
```
Comment (jmuehlner) [inline src/libguac/guacamole/user-fntypes.h]: Missing parameter comments

File: src/libguac/parser.c [hunk 36]
```diff
@@ -31,46 +31,49 @@
...
+    if (parser == NULL) {
         guac_error = GUAC_STATUS_NO_MEMORY;
-        guac_error_message = "Insufficient memory to allocate instruction";
+        guac_error_message = "Insufficient memory to allocate parser";
         return NULL;
     }
 
-    guac_instruction_reset(instruction);
-    return instruction;
+    /* Init parse start/end marksers */
```
Comment (jmuehlner) [inline src/libguac/parser.c]: Marskers

File: src/libguac/socket-fd.c [hunk 37]
```diff
@@ -61,32 +147,140 @@ ssize_t __guac_socket_fd_read_handler(guac_socket* socket,
...
-        guac_error = GUAC_STATUS_SEE_ERRNO;
-        guac_error_message = "Error writing data to socket";
     }
 
+    /* All bytes have been written, possibly some to the internal buffer */
+    return original_count;
+
+}
+
+static ssize_t guac_socket_fd_write_handler(guac_socket* socket,
```
Comment (jmuehlner) [inline src/libguac/socket-fd.c]: Missing comment

File: src/libguac/socket-fd.c [hunk 38]
```diff
@@ -61,32 +147,140 @@ ssize_t __guac_socket_fd_read_handler(guac_socket* socket,
...
+            return 1;
+
+        data->written = 0;
+    }
+
+    return 0;
+
+}
+
+static ssize_t guac_socket_fd_flush_handler(guac_socket* socket) {
```
Comment (jmuehlner) [inline src/libguac/socket-fd.c]: Missing comment

File: src/libguac/socket-fd.c [hunk 39]
```diff
@@ -61,32 +147,140 @@ ssize_t __guac_socket_fd_read_handler(guac_socket* socket,
...
+
+    /* Relinquish exclusive access to buffer */
+    pthread_mutex_unlock(&(data->buffer_lock));
+
     return retval;
+
 }
 
-int __guac_socket_fd_select_handler(guac_socket* socket, int usec_timeout) {
+static int guac_socket_fd_select_handler(guac_socket* socket,
```
Comment (jmuehlner) [inline src/libguac/socket-fd.c]: Missing comment

File: src/libguac/socket-nest.c [hunk 40]
```diff
@@ -105,6 +105,16 @@ ssize_t __guac_socket_nest_write_handler(guac_socket* socket,
 
 }
 
+static int __guac_socket_nest_free_handler(guac_socket* socket) {
```
Comment (jmuehlner) [inline src/libguac/socket-nest.c]: Missing comment

File: src/libguac/user-handlers.c [hunk 41]
```diff
@@ -0,0 +1,448 @@
...
+    if (user->key_handler)
+        return user->key_handler(
+            user,
+            atoi(argv[0]), /* keysym */
+            atoi(argv[1])  /* pressed */
+        );
+    return 0;
+}
+
+static guac_stream* __get_input_stream(guac_user* user, int stream_index) {
```
Comment (jmuehlner) [inline src/libguac/user-handlers.c]: Missing comment

File: src/libguac/user-handlers.c [hunk 42]
```diff
@@ -0,0 +1,448 @@
...
+        guac_protocol_send_ack(user->socket, &dummy_stream,
+                "Invalid stream index", GUAC_PROTOCOL_STATUS_CLIENT_BAD_REQUEST);
+        return NULL;
+    }
+
+    return &(user->__input_streams[stream_index]);
+
+}
+
+static guac_stream* __get_open_input_stream(guac_user* user, int stream_index) {
```
Comment (jmuehlner) [inline src/libguac/user-handlers.c]: Missing comment

File: src/libguac/user-handlers.c [hunk 43]
```diff
@@ -0,0 +1,448 @@
...
+        guac_protocol_send_ack(user->socket, &dummy_stream,
+                "Invalid stream index", GUAC_PROTOCOL_STATUS_CLIENT_BAD_REQUEST);
+        return NULL;
+    }
+
+    return stream;
+
+}
+
+static guac_stream* __init_input_stream(guac_user* user, int stream_index) {
```
Comment (jmuehlner) [inline src/libguac/user-handlers.c]: Missing comment

File: src/libguac/user-handlers.h [hunk 44]
```diff
@@ -21,26 +21,26 @@
...
 
 #include "client.h"
-#include "instruction.h"
+#include "timestamp.h"
 
 /**
  * Internal handler for Guacamole instructions.
  */
-typedef int __guac_instruction_handler(guac_client* client, guac_instruction* copied);
+typedef int __guac_instruction_handler(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 45]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
  * is received, this handler will be called. Sync instructions are automatically
  * handled, thus there is no client handler for sync instruction.
  */
-int __guac_handle_sync(guac_client* client, guac_instruction* instruction);
+int __guac_handle_sync(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 46]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the mouse instruction. When a mouse instruction
  * is received, this handler will be called. The client's mouse handler will
  * be invoked if defined.
  */
-int __guac_handle_mouse(guac_client* client, guac_instruction* instruction);
+int __guac_handle_mouse(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 47]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the key instruction. When a key instruction
  * is received, this handler will be called. The client's key handler will
  * be invoked if defined.
  */
-int __guac_handle_key(guac_client* client, guac_instruction* instruction);
+int __guac_handle_key(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 48]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the clipboard instruction. When a clipboard instruction
  * is received, this handler will be called. The client's clipboard handler will
  * be invoked if defined.
  */
-int __guac_handle_clipboard(guac_client* client, guac_instruction* instruction);
+int __guac_handle_clipboard(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 49]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the file instruction. When a file instruction
  * is received, this handler will be called. The client's file handler will
  * be invoked if defined.
  */
-int __guac_handle_file(guac_client* client, guac_instruction* instruction);
+int __guac_handle_file(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 50]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the pipe instruction. When a pipe instruction
  * is received, this handler will be called. The client's pipe handler will
  * be invoked if defined.
  */
-int __guac_handle_pipe(guac_client* client, guac_instruction* instruction);
+int __guac_handle_pipe(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 51]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the ack instruction. When a ack instruction
  * is received, this handler will be called. The client's ack handler will
  * be invoked if defined.
  */
-int __guac_handle_ack(guac_client* client, guac_instruction* instruction);
+int __guac_handle_ack(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 52]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the blob instruction. When a blob instruction
  * is received, this handler will be called. The client's blob handler will
  * be invoked if defined.
  */
-int __guac_handle_blob(guac_client* client, guac_instruction* instruction);
+int __guac_handle_blob(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 53]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the end instruction. When a end instruction
  * is received, this handler will be called. The client's end handler will
  * be invoked if defined.
  */
-int __guac_handle_end(guac_client* client, guac_instruction* instruction);
+int __guac_handle_end(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 54]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the get instruction. When a get instruction
  * is received, this handler will be called. The client's get handler will
  * be invoked if defined.
  */
-int __guac_handle_get(guac_client* client, guac_instruction* instruction);
+int __guac_handle_get(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 55]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the put instruction. When a put instruction
  * is received, this handler will be called. The client's put handler will
  * be invoked if defined.
  */
-int __guac_handle_put(guac_client* client, guac_instruction* instruction);
+int __guac_handle_put(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 56]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the size instruction. When a size instruction
  * is received, this handler will be called. The client's size handler will
  * be invoked if defined.
  */
-int __guac_handle_size(guac_client* client, guac_instruction* instruction);
+int __guac_handle_size(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

File: src/libguac/user-handlers.h [hunk 57]
```diff
@@ -64,91 +64,91 @@ typedef struct __guac_instruction_handler_mapping {
...
 
 /**
  * Internal initial handler for the disconnect instruction. When a disconnect instruction
  * is received, this handler will be called. Disconnect instructions are automatically
  * handled, thus there is no client handler for disconnect instruction.
  */
-int __guac_handle_disconnect(guac_client* client, guac_instruction* instruction);
+int __guac_handle_disconnect(guac_user* user, int argc, char** argv);
```
Comment (jmuehlner) [inline src/libguac/user-handlers.h]: Missing parameter comments

**Comment by mike-jumper** [comment]
Gah. Finally.

**Pull request merged**